apps/block_scout_web/lib/block_scout_web/controllers/verified_contracts_controller.ex

```diff
@@ -2,17 +2,160 @@ defmodule BlockScoutWeb.VerifiedContractsController do
+      Keyword.merge(
+        [contracts: contracts],
+        prepare_paging_vars(
+          conn,
+          paging_options,
+          params,
+          contract_count,
+          &path_for_paging_options/2
+        )
+      )
+    )
   end
 
-  def get_verified_contracts() do
-    Explorer.Chain.SmartContract
+  defp get_verified_contracts(paging_options, filter) do
+    Contract
     |> preload(:address)
+    |> handle_filter(filter)
+    |> handle_paging_options(paging_options)
     |> Explorer.Repo.all()
   end
-end
\ No newline at end of file
+
+  defp get_verified_contract_count(filter) do
+    Contract
+    |> handle_filter(filter)
+    |> Explorer.Repo.aggregate(:count, :id)
+  end
+
+  defp handle_paging_options(query, paging_options) do
+    offset = (paging_options.page_number - 1) * paging_options.page_size
+
+    query
+    |> limit(^paging_options.page_size)
+    |> offset(^offset)
+    |> order_by(asc: :name) # TODO parametrize (use fragment)
+  end
+
+  defp handle_filter(query, filter) do
+    if not is_nil(filter)do
+      # TODO implement WHERE clause
+      query
+    else
+      query
+    end
+  end
+
+  defp path_for_paging_options(conn, paging_options) do
+    verified_contracts_path(conn, :index, paging_options)
+  end
+
+  # TODO below functions should be moved to a helper
+
+  # TODO validate paging options
+  # TODO add error handling
+  defp extract_paging_options(
+    params,
+    total_item_count,
+    allowed_order_fields
+  ) do
+    page_size = if Map.has_key?(params, "page_size") do
+      case Integer.parse(Map.get(params, "page_size", @default_page_size)) do
+        {page_size, _} -> page_size
+        :error -> @default_page_size
+      end
+    else
+      @default_page_size
+    end
+    page_number = if Map.has_key?(params, "page_number") do
+      page_number = case Integer.parse(Map.get(params, "page_number", 1)) do
+        {page_number, _} -> page_number
+        :error -> 1
+      end
+
+      if page_number > ceil(total_item_count / page_size) do
+        1
+      else
+        page_number
+      end
+    else
+      1
+    end
+    order_field = if Enum.count(allowed_order_fields) > 0 do
+      if Map.has_key?(params, "order_field") and Enum.member?(allowed_order_fields, Map.get(params, "order_field")) do
+        Map.get(params, "order_field")
+      else
+        Enum.at(allowed_order_fields, 0)
+      end
+    else
+      nil
+    end
+    order_dir = if Map.has_key?(params, "order_dir") and Enum.member?(["desc", "asc"], Map.get(params, "order_dir")) do
+      Map.get(params, "order_dir")
+    else
+      "asc"
+    end
+
+    %PagingOptions{
+      order_field: order_field,
+      order_dir: order_dir,
+      page_size: page_size,
+      page_number: page_number,
+    }
+  end
+
+  defp prepare_paging_vars(
+    conn,
+    paging_options,
+    params,
+    total_item_count,
+    path_fun
+  ) do
+    last_page_number = ceil(total_item_count / paging_options.page_size)
+
+    [
```

Comment:
This is pretty difficult to read with so many conditionals and indirection in the declaration of a list literal, but I would personally put this logic in the template directly - calculate the path for the next page to be show directly on render of the template, rather than in the controller directly. 

This feels like view / presentation logic elevated to the controller level.

Reply:
As for the readability I totally agree, was thinking about extracting parts of it into separate functions. Would a view helper be a more appropriate place for it?